gh-90548: Make musl test skips smarter (fixes Alpine errors)

bitdancer opened the pull request:
Make musl test skips smarter (fixes Alpine errors)

A relatively small number of tests fail when the underlying c library is
provided by musl.  This was originally reported in bpo-46390 by
Christian Heimes.  Among other changes, these tests were marked for
skipping in gh-31947/ef1327e3 as part of bpo-40280 (emscripten support),
but the skips were conditioned on the *platform* being emscripten (or
wasi, skips for which ere added in 9b50585e02).

In gh-131071 Victor Stinner added a linked_to_musl function to enable
skipping a test in test_math that fails under musl, like it does on a
number of other platforms.  This check can successfully detect that
python is running under musl on Alpine, which was the original problem
report in bpo-46390.

This PR replaces Victor's solution with an enhancement to
platform.libc_ver that does the check more cheaply, and also gets the
version number.  The latter is important because the math test being
skipped is due to a bug in musl that has been fixed, but as of this
checkin date has not yet been released.  When it is, the test skip can
be fixed to check for the minimum needed version.

The enhanced version of linked_to_musl is also used to do the skips of
the other tests that generically fail under musl, as opposed to
emscripten or wasi only failures.  This will allow these tests to be
skipped automatically on Alpine.

This PR does *not* enhance libc_ver to support emscripten and wasi, as
I'm not familiar with those platforms; instead it returns a version
triple of (0, 0, 0) for those platforms.  This means the musl tests will
be skipped regardless of musl version, so ideally someone will add
support to libc_ver for these platforms.


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--131313.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

<!-- gh-issue-number: gh-90548 -->
* Issue: gh-90548
<!-- /gh-issue-number -->

bitdancer committed:
Make musl test skips smarter (fixes Alpine errors)

A relatively small number of tests fail when the underlying c library is
provided by musl.  This was originally reported in bpo-46390 by
Christian Heimes.  Among other changes, these tests were marked for
skipping in gh-31947/ef1327e3 as part of bpo-40280 (emscripten support),
but the skips were conditioned on the *platform* being emscripten (or
wasi, skips for which ere added in 9b50585e02).

In gh-131071 Victor Stinner added a linked_to_musl function to enable
skipping a test in test_math that fails under musl, like it does on a
number of other platforms.  This check can successfully detect that
python is running under musl on Alpine, which was the original problem
report in bpo-46390.

This PR replaces Victor's solution with an enhancement to
platform.libc_ver that does the check more cheaply, and also gets the
version number.  The latter is important because the math test being
skipped is due to a bug in musl that has been fixed, but as of this
checkin date has not yet been released.  When it is, the test skip can
be fixed to check for the minimum needed version.

The enhanced version of linked_to_musl is also used to do the skips of
the other tests that generically fail under musl, as opposed to
emscripten or wasi only failures.  This will allow these tests to be
skipped automatically on Alpine.

This PR does *not* enhance libc_ver to support emscripten and wasi, as
I'm not familiar with those platforms; instead it returns a version
triple of (0, 0, 0) for those platforms.  This means the musl tests will
be skipped regardless of musl version, so ideally someone will add
support to libc_ver for these platforms.

bitdancer committed:
Platform tests and bug fixes.

In adding tests for the new platform code I found a bug in the old code:
if a valid version is passed for version and it is greater than the
version found for an so *and* there is no glibc version, then the
version from the argument was returned.  The code changes here fix
that, as well as fixing my own broken additions.

bitdancer committed:
Add docs, including some existing is_xxx's.

vstinner commented on Doc/library/test.rst:
```diff
@@ -831,6 +851,15 @@ The :mod:`test.support` module defines the following functions:
+   ``musl``, otherwise return a version triple, either ``(0, 0, 0)`` if the
+   version is unknown, or the actual version if it is known.  Intended for use
+   in ``skip`` decorators.  ``emscripten`` and ``wasi`` are assumed to be
+   compiled with ``musl``; othewise ``platform.libc_ver`` is checked.
```
```suggestion
   compiled with ``musl``; otherwise ``platform.libc_ver`` is checked.
```

vstinner commented on Lib/test/test_math.py:
```diff
@@ -2772,6 +2772,9 @@ def test_fma_infinities(self):
         or (sys.platform == "android" and platform.machine() == "x86_64")
         or support.linked_to_musl(),  # gh-131032
         f"this platform doesn't implement IEE 754-2008 properly")
+    # XXX musl is fixed but the fix is not yet released; when the fixed version
```
```suggestion
    # gh-131032: musl is fixed but the fix is not yet released; when the fixed version
```

vstinner commented on Lib/test/test_os.py:
```diff
@@ -2555,15 +2555,18 @@ def test_fchown(self):
+        support.linked_to_musl(),
+        'musl pathconf ignores the file descriptor and returns a constant',
+        )
+    def test_fpathconf_with_bad_fd(self):
```
```suggestion
    def test_fpathconf_bad_fd(self):
```

vstinner commented on Lib/test/test_platform.py:
```diff
@@ -552,6 +552,7 @@ def test_libc_ver(self):
                 (b'GLIBC_2.9', ('glibc', '2.9')),
                 (b'libc.so.1.2.5', ('libc', '1.2.5')),
                 (b'libc_pthread.so.1.2.5', ('libc', '1.2.5_pthread')),
+                (b'/aports/main/musl/src/musl-1.2.5', ('musl', '1.2.5')),
```
Can you add tests on version with 2 numbers and 4 numbers? Like `/aports/main/musl/src/musl-1.2` and `/aports/main/musl/src/musl-1.2.5.1`.

bitdancer replied:
I could, but I don't believe musl will ever have such release numbers, so is it worth doing?

vstinner replied:
Parsing a binary file using a regular expression is a strange thing for me. So I prefer to have tests to make sure that the implementation is reliable. The question is more if the regex will match versions with 2 members (x.y) or 4 members (x.y.z.a), or if it only matchs version with 3 members (x.y.z).

bitdancer replied:
Fair enough, though the existing tests don't do that, nor do they test the no-match cases.  I could add such tests, but at that point I'd want to rewrite this test method to split it up into multiple tests, and I don't think that's in scope for this PR :)  I'll add the couple of extra checks, though.

vstinner commented on Lib/test/test_support.py:
```diff
@@ -746,7 +746,18 @@ def test_get_signal_name(self):
+        if support.is_wasi:
+            self.assertTrue(linked)
+        # The value is cached, so make sure it returns the same value again.
+        self.assertEqual(linked, support.linked_to_musl())
```
```suggestion
        self.assertIs(linked, support.linked_to_musl())
```

vstinner commented on Lib/test/test_support.py:
```diff
@@ -746,7 +746,18 @@ def test_get_signal_name(self):
         linked = support.linked_to_musl()
-        self.assertIsInstance(linked, bool)
+        self.assertIsNotNone(linked)
+        if support.is_wasi:
```
What about is_emscripten?

bitdancer replied:
I used is_wasi as an example, but it doesn't hurt to add emscripten as well.

vstinner commented on Lib/test/test_support.py:
```diff
@@ -746,7 +746,18 @@ def test_get_signal_name(self):
-        self.assertIsInstance(linked, bool)
+        self.assertIsNotNone(linked)
+        if support.is_wasi:
+            self.assertTrue(linked)
```
Check maybe even that it's equal to (0, 0, 0)?

bitdancer replied:
I'd rather not...eventually I hope it will be a version number.

picnixz commented on Lib/platform.py:
```diff
@@ -189,22 +189,28 @@ def libc_ver(executable=None, lib='', version='', chunksize=16384):
+        br'(libc(_\w+)?\.so(?:\.(\d[0-9.]*))?)'
+        br'|'
+        br'(musl-([0-9.]+))'
+        br'',
```
Is the empty string at the end intended?

bitdancer replied:
Yes.  It will produce cleaner diffs if more clauses are added in the future (equivalent to having an unneeded comma on the last line of a multiline list).

picnixz commented on Lib/test/support/__init__.py:
```diff
@@ -3017,20 +3017,31 @@ def is_libssl_fips_mode():
+    # emscripten (at least as far as we're concerned) and wasi use musl,
+    # but platform doesn't know how to get the version, so set it to zero.
+    if is_emscripten or is_wasi:
+        return (_linked_to_musl := (0, 0, 0))
```
TIL that we can also do this to save a few lines... Not sure if it's the most readable but I don't mind it.

bitdancer replied:
Yeah, I wasn't quite sure about it, but it seemed like a reasonable use of the operator.  I'm fine with being more wordy if that is considered better style.

picnixz commented on Lib/test/test_os.py:
```diff
@@ -2555,15 +2555,18 @@ def test_fchown(self):
+    @unittest.skipIf(
+        support.linked_to_musl(),
+        'musl pathconf ignores the file descriptor and returns a constant',
+        )
```
```suggestion
    )
```

bitdancer replied:
I prefer the style I used because it lines up better with how python's statements work (first line only is outdented, lack of outdent signals return to outer logic block).  This is accepted by PEP 8, are there additional style constrains on the cpython code base these days?

picnixz replied:
I haven't seen lots of use of hanging indents in the lib. There is no additional constraints but usually we try to be consistent with the surrounding code when possible.

picnixz commented on Lib/test/test_platform.py:
```diff
@@ -563,14 +564,29 @@ def test_libc_ver(self):
+                # test match at chunk boundary
+                f.write(b'x'*(chunksize - 10))
+                f.write(data)
+            self.assertEqual(
+                expected,
+                platform.libc_ver(
+                    filename,
+                    lib='test',
+                    version='100.1.0',
+                    chunksize=chunksize,
+                    ),
+                )
```
```suggestion
            self.assertEqual(
                expected,
                platform.libc_ver(
                    filename,
                    lib='test',
                    version='100.1.0',
                    chunksize=chunksize,
                ),
            )
```

picnixz commented on Lib/test/test_support.py:
```diff
@@ -746,7 +746,18 @@ def test_get_signal_name(self):
+        # The unlike libc, the musl version is a triple.
+        if linked:
+            self.assertIsInstance(linked, tuple)
+            self.assertEqual(3, len(linked))
```
```suggestion
            self.assertEqual(len(linked), 3)
```

bitdancer replied:
Expected should come first because it makes the assertEqual diffs clearer in the general case.  In this case it wouldn't matter, but I prefer to be consistent.

picnixz replied:
Ah, most of the tests I wrote and read were actually swapped. I think it's half-half so I don't mind.

picnixz commented on Lib/test/test_platform.py:
```diff
@@ -563,14 +564,29 @@ def test_libc_ver(self):
+        # make sure that eg 1.9 is seen as older than 1.23.4, and that
+        # the arguments don't count even if they are set.
+        chunksize = 200
+        for data, expected in (
+                (b'GLIBC_1.23.4\0GLIBC_1.9\0GLIBC_1.21\0', ('glibc', '1.23.4')),
+                (b'libc.so.2.4\0libc.so.9\0libc.so.23.1\0', ('libc', '23.1')),
+                (b'musl-1.4.1\0musl-2.1.1\0musl-2.0.1\0', ('musl', '2.1.1')),
+                (b'no match here, so defaults are used', ('test', '100.1.0')),
+            ):
```
```suggestion
        for data, expected in (
            (b'GLIBC_1.23.4\0GLIBC_1.9\0GLIBC_1.21\0', ('glibc', '1.23.4')),
            (b'libc.so.2.4\0libc.so.9\0libc.so.23.1\0', ('libc', '23.1')),
            (b'musl-1.4.1\0musl-2.1.1\0musl-2.0.1\0', ('musl', '2.1.1')),
            (b'no match here, so defaults are used', ('test', '100.1.0')),
        ):
```

bitdancer replied:
I believe my version is more consistent with pep8, though you could argue that the `):` should be indented four more spaces as well.

picnixz replied:
Up to you. I won't block the PR for cosmetics only!

picnixz reviewed (commented):
For my suggestions with dedents, please check them manually as I'm not entirely sure that I didn't mix tabs and spaces.

bitdancer committed:
Address vstinner review comments.

bitdancer committed:
Test some version number variants we accept.

bitdancer committed:
Add news item about libc_ver enhancement.

vstinner commented on Lib/platform.py:
```diff
@@ -189,22 +189,28 @@ def libc_ver(executable=None, lib='', version='', chunksize=16384):
         pos = 0
         while pos < len(binary):
-            if b'libc' in binary or b'GLIBC' in binary:
+            if b'libc' in binary or b'GLIBC' or 'musl' in binary:
```
```suggestion
            if b'libc' in binary or b'GLIBC' or b'musl' in binary:
```

bitdancer replied:
Oh, that's a nasty one.  The actual fix is `b'GLIBC` in binary or b'musl' in binary`.  That's why it passed without the b...it just ran slower ;)  (A lot slower, somewhat to my surprise.)

vstinner reviewed (approved):
LGTM, but please fix the important typo `'musl' in binary`.

picnixz reviewed (approved):
As I said, I won't block the PR for cosmetics only though I would have aligned with the surrounding/existing code when possible.

serhiy-storchaka commented on Lib/platform.py:
```diff
@@ -189,22 +189,28 @@ def libc_ver(executable=None, lib='', version='', chunksize=16384):
+        br'|'
+        br'(musl-([0-9.]+))'
+        br'',
+        re.ASCII)
```
BTW, we can switch to the VERBOSE mode, which would be slightly more clear.
```suggestion
    libc_search = re.compile(br'''
          (__libc_init)
        | (GLIBC_([0-9.]+))
        | (libc(_\w+)?\.so(?:\.(\d[0-9.]*))?)
        | (musl-([0-9.]+))
        ''',
        re.ASCII | re.VERBOSE)
```

bitdancer replied:
I like that.

serhiy-storchaka commented on Lib/test/support/__init__.py:
```diff
@@ -3017,20 +3017,31 @@ def is_libssl_fips_mode():
+    # emscripten (at least as far as we're concerned) and wasi use musl,
+    # but platform doesn't know how to get the version, so set it to zero.
+    if is_emscripten or is_wasi:
+        return (_linked_to_musl := (0, 0, 0))
```
I do not think that it is necessary to use the `:=` operator here.

bitdancer replied:
It is never *necessary* to use the `:=` operator, so do you mean you prefer the more verbose style (which I'm fine with changing to), or am I missing something?

serhiy-storchaka replied:
I prefer the more verbose style, which is also more sequential and clear.

bitdancer replied:
Done.

serhiy-storchaka commented on Lib/test/test_strptime.py:
```diff
@@ -544,10 +544,7 @@ def test_date_locale(self):
-        "musl libc issue on Emscripten, bpo-46390"
-    )
+    @unittest.skipIf(support.linked_to_musl(), "musl libc issue, bpo-46390")
     @run_with_locales('LC_TIME', 'en_US', 'fr_FR', 'de_DE', 'ja_JP',
```
We can add the `linked_to_musl()` check in `run_with_locales()` and `run_with_locale()`. This may be non-trivial, depending on arguments, a dummy run can be supported even if no suitable locale was found. For now, the proposed change is good as is.

serhiy-storchaka reviewed (approved):
LGTM in general.

bitdancer committed:
Fix 'in' phrase logic.

This didn't cause a test failure because the broken logic caused the
regex to always run...which is a lot slower.

bitdancer committed:
Prettify re expression using re.VERBOSE.

bitdancer committed:
Use more verbose style rather than :=
